pkg/cache/v3/snapshot.go

```diff
@@ -81,19 +82,47 @@ func (s *Snapshot) Consistent() error {
 	if s == nil {
 		return errors.New("nil snapshot")
 	}
-	endpoints := GetResourceReferences(s.Resources[types.Cluster].Items)
-	if len(endpoints) != len(s.Resources[types.Endpoint].Items) {
-		return fmt.Errorf("mismatched endpoint reference and resource lengths: %v != %d", endpoints, len(s.Resources[types.Endpoint].Items))
-	}
-	if err := superset(endpoints, s.Resources[types.Endpoint].Items); err != nil {
-		return err
+
+	consolidated := make(map[resource.Type]map[string]bool)
+	// References made by CDS.
+	GetResourceReferencesByReference(s.Resources[types.Cluster].Items, consolidated)
+	// References made by LDS.
+	GetResourceReferencesByReference(s.Resources[types.Listener].Items, consolidated)
+	// References made by SRDS.
+	GetResourceReferencesByReference(s.Resources[types.ScopedRoute].Items, consolidated)
+
+	// Loop through each referenced resource.
+	referencedResponseTypes := map[types.ResponseType]bool{
+		types.Endpoint:    true,
+		types.ScopedRoute: true,
+		types.Route:       true,
 	}
 
-	routes := GetResourceReferences(s.Resources[types.Listener].Items)
-	if len(routes) != len(s.Resources[types.Route].Items) {
-		return fmt.Errorf("mismatched route reference and resource lengths: %v != %d", routes, len(s.Resources[types.Route].Items))
+	for idx, items := range s.Resources {
+
+		// We only want to check resource types that are expected to be referenced by another resource type.
+		// Basically, if the consistency relationship is modeled as a DAG, we only want
+		// to check nodes that are expected to have edges pointing to it.
+		responseType := types.ResponseType(idx)
+		if _, ok := referencedResponseTypes[responseType]; ok {
```

Comment:
Don't you also need to check the returned value from this map access? The second value only checks to see if the key in the map exists so `types.Endpoint` could be `false` and `ok` will be true.

```
if yes, ok := referencedResponseTypes[responseType]; ok && yes {

}
```

Reply:
This map is supposed to be a "set", but I see the concern. I'm going to change the struct type to map[string]struct{} that way there is no confusion with the boolean values.